### ISSUE TITLE
Proposal: Custom API endpoint for client side lock verification

### ISSUE THREAD

**Issue by technoweenie** [git-lfs/git-lfs#1891]
Git LFS in `master` verifies that pushes are not updating locked files based on work in #1815.

1. Call [`GET /locks`](https://github.com/git-lfs/git-lfs/blob/e029224251e2d8b134ee92f3cc00e001fdcbfb40/docs/proposals/locking_api.md#get-locks) to get all locks from the server.
2. [Sort locks into `owned` and `unowned` buckets](https://github.com/git-lfs/git-lfs/pull/1815/files#diff-3cff4ed9c5b60b5b81240c88cd74e404R98) based on the `user.name` and `user.email` git config values.

While this is fine for a prototype, it's not good enough for production.

My proposal is to introduce a new endpoint specifically for these client side checks:

```
> POST /locks/verify
> Accept: application/vnd.git-lfs+json
> Authorization: Basic
> Content-Type: application/vnd.git-lfs+json
>
> {
> }

< HTTP/1.1 200 Ok
< Content-Type: application/vnd.git-lfs+json
<
< {
<   "ours": [
<     // full "lock" properties omitted to save space
<     { "id": "1", "path": "foo/bar" }
<   ],
<   "theirs": [
<     { "id": "2", "path": "foo/baz" }
<   ],
<   next_cursor: "optional-next-id",
<   error: "optional error"
< }
```

This is the bare minimum that could replace the existing code. It leaves us a lot of flexibility to add properties in the future. Maybe we need to send other details in the request to the server to support multi-branch workflows, for example.

The immediate benefit here is that it puts the burden of sorting the locks into owned/unowned buckets is on the server, which is able to verify user credentials.

Also, any LFS server that does this validation on the server can return a `410 Gone` response for just this `/locks/verify` endpoint. This would disable client side lock verification if it's unnecessary, based on this proposal: Locking Pre-Push hook should be optional #1884.

**Comment by ttaylorr** [comment]
> The immediate benefit here is that it puts the burden of sorting the locks into owned/unowned buckets is on the server, which is able to verify user credentials.

Definitely on board with this -- offloading that work to an authoritative source of data is a huge win. Here are a few of my thoughts:

Some high-level questions:

1. Really happy that we have opportunity to grow with this endpoint. I think this is the right approach to be taking.

Some implementation questions:

1. How does the `next_cursor` property work? Since lock IDs are `AUTO INCREMENT`, does a non-zero `next_cursor` indicate that all the locks in the two buckets returned have IDs less than `next_cursor`?
2. Any locks in the negative intersection of `required_locks` and `your_locks` halts the push. Since all of the data in `your_locks` will already be present in `required_locks`, do you think we could send an array of lock IDs instead of the full lock metadata?

**Comment by technoweenie** [comment]
> How does the next_cursor property work? Since lock IDs are AUTO INCREMENT, does a non-zero next_cursor indicate that all the locks in the two buckets returned have IDs less than next_cursor?

I envisioned this working exactly the same as `GET /locks`. Though instead of returning a single `locks` array of objects, the server will split them into `required_locks` and `your_locks`.

> Any locks in the negative intersection of required_locks and your_locks halts the push. Since all of the data in your_locks will already be present in required_locks, do you think we could send an array of lock IDs instead of the full lock metadata?

Perhaps the key names `required_locks` and `your_locks` are not good... but my intention was that essentially the server is doing the partitioning that [the pre-push hook does now](https://github.com/git-lfs/git-lfs/blob/e029224251e2d8b134ee92f3cc00e001fdcbfb40/commands/uploader.go#L100-L109). The example had some copy pasta, so I changed the ID of the lock in `your_locks` so this is more apparent.

**Comment by ttaylorr** [comment]
> Perhaps the key names `required_locks` and `your_locks` are not good... but my intention was that essentially the server is doing the partitioning that [the pre-push hook does now](https://github.com/git-lfs/git-lfs/blob/e029224251e2d8b134ee92f3cc00e001fdcbfb40/commands/uploader.go#L100-L109).

Ah, I misread this as meaning:

- `required_locks` - these are all of the active locks
- `your_locks` - this is the subset of locks that you own

so a non-zero intersection of the two would mean that there are `required_locks` you don't own. I was thinking a bit more about this this morning and thought that an appropriate set of names might be "ours" and "theirs", which is terminology that Git uses.

Ours would be the set of locks that is owned by the authorized user, and theirs would be everything else.

I think the pagination semantics are fine. 👍

**Comment by technoweenie** [comment]
> I was thinking a bit more about this this morning and thought that an appropriate set of names might be "ours" and "theirs", which is terminology that Git uses.

:+1:

**Comment by sinbad** [comment]
👍 delegating to the server to decide identity is a good move.

**Comment by technoweenie** [comment]
Implemented in #1905 :metal:  Thanks for the feedback.
